includes/class-wpcom-legacy-redirector-ui.php

```diff
@@ -0,0 +1,429 @@
+						$parent_slug = $parent->post_name;
+						echo '/' . $parent_slug;
+						$redirect = home_url() . '/' . $parent_slug;
+					}
+				}
+				$status = $this->check_if_404( $redirect );
+
+				// Check if $redirect is invalid.
+				if ( ! wp_validate_redirect( $redirect, false ) ) {
+					wp_safe_redirect( add_query_arg(
+						array(
+							'validate'  => 'invalid',
+							'ids'       => $post->ID,
+						), $sendback
+					) );
+					exit();
+				}
+				// Check if $redirect is a 404.
+				if ( 404 === $status ) {
+					wp_safe_redirect( add_query_arg(
+						array(
+							'validate'  => '404',
+							'ids'       => $post->ID,
+						), $sendback
+					) );
+					exit();
+				}
+				// Check if $redirect is not publicly visible.
+				if ( 'private' === $redirect ) {
+					wp_safe_redirect( add_query_arg(
+						array(
+							'validate'  => 'private',
+							'ids'       => $post->ID,
+						), $sendback
+					) );
+					exit();
+				}
+				// Check if $redirect is pointing to a null Post ID.
+				if ( 'null' === $redirect ) {
+					wp_safe_redirect( add_query_arg(
+						array(
+							'validate'  => 'null',
+							'ids'       => $post->ID,
+						), $sendback
+					) );
+					exit();
+				}
+				// Check if $redirect is valid.
+				if ( wp_validate_redirect( $redirect, false ) && 404 !== $status || 'valid' === $redirect ) {
+					wp_safe_redirect( add_query_arg(
+						array(
+							'validate'  => 'valid',
+							'ids'       => $post->ID,
+						), $sendback
+					) );
+					exit();
+				}
+			}
+		}
+	}
+	public function generate_page_html() {
+		if ( ! current_user_can( 'manage_redirects' ) ) {
+			return;
+		}
+
+		$errors   = array();
+		$messages = array();
+		if ( class_exists( 'WPCOM_Legacy_Redirector' ) ) {
+			if ( isset( $_POST['from_url'] ) && isset( $_POST['redirect_to'] ) ) {
+				if (
+					! isset( $_POST['redirect_nonce_field'] )
+					|| ! wp_verify_nonce( $_POST['redirect_nonce_field'] )
+				) {
+					$errors[] = array(
+						'label'   => __( 'Error', 'wpcom-legacy-redirector' ),
+						'message' => __( 'Sorry, your nonce did not verify.', 'wpcom-legacy-redirector' ),
+					);
+				} else {
+					$from_url	= sanitize_text_field( $_POST['from_url'] );
+					$redirect_to = sanitize_text_field( $_POST['redirect_to'] );
+
+					// Check if $from_url is not a 404.  Only 404 links are redirected in this plugin.
+					$status = $this->check_if_404( home_url() . $from_url );
+					if ( 404 !== $status ) {
+						$from_url = '';
+					}
+
+					if ( WPCOM_Legacy_Redirector::validate( $from_url, $redirect_to ) ) {
+						$output = WPCOM_Legacy_Redirector::insert_legacy_redirect( $from_url, $redirect_to, false );
+						if ( true === $output ) {
+							$link	   = '<a href="' . esc_url( $from_url ) . '" target="_blank">' . esc_url( $from_url ) . '</a>';
+							$messages[] = __( 'The redirect was added successfully. Check Redirect: ', 'wpcom-legacy-redirector' ) . $link;
+						} else {
+							if ( false === $output ) {
```

Comment:
Under what circumstances does it return false?
Under which does it return an error message?
Why not always return an error message instead of sometimes false and sometimes WP_Error?